### PR TITLE
Add support for fmsub and fnmsub instructions on x86

### PR DESCRIPTION
**Pull request by UnlimitedHummus**
Issue #8602

Allows generating fmsub and fnmsub instructions as mentioned in the issue.

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/lower.isle [hunk 1]
```diff
@@ -2860,6 +2860,38 @@
...
+(rule 2 (lower (has_type ty (fma x y (fneg z))))
+      (if-let $true (use_fma))
+      (fmsub ty x y z))
+
+;; fmsub and fnmsub
+(decl fmsub (Type Value Value Value) Xmm)
+(decl fnmsub (Type Value Value Value) Xmm)
+
+;; Base case, will sink a load of `z` automatically.
+(rule (fmsub ty x y z) (x64_vfmsub213 ty x y z))
```
Comment (fitzgen) [inline cranelift/codegen/src/isa/x64/lower.isle:2873]: I forget what the details of our lowering and ISLE conversions are these days, but does this need to guard on sinkable loads? Or will it force things to registers as written? (In which case the comment should be updated)

cc @jameysharp, who I think would know the answer to this question.

**Review by fitzgen — COMMENTED** [review]
Thanks! Overall looks good to me, modulo the one comment below.

Can you also add some run tests in addition to the compile tests, to make sure that the runtime behavior is as-expected as well? Since FMA is not quite the same across architectures, it may make sense to put these run tests under `cranelift/filetests/filetests/isa/x64` as well.

File: cranelift/codegen/src/isa/x64/lower.isle (hunk 1 above)
Comment (jameysharp) [reply cranelift/codegen/src/isa/x64/lower.isle:2873]: I'm not actually sure. But I see that this is identical to the existing lowering rules above aside from replacing `add` with `sub` and so on. So it's probably fine?

Comment (UnlimitedHummus) [reply cranelift/codegen/src/isa/x64/lower.isle:2873]: I did just work off of the rules for the `add` instructions. However, its probably related to the semantics of `z` matching into `XmmMem` in the rule for `x64vfmsub213`. I didn't look much into how that works, but the first two test cases actually test this rule both in the case of passing a value in a register or a sinkable load. There, the load is definitely sunk.

**Comment by UnlimitedHummus** [comment]
Sure, I'll add the runtime tests later, when I have time.

**Comment by afonso360** [comment]
We already have some run tests that run on all arch's here ([`1`](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/filetests/filetests/runtests/fma.clif), [`2`](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/filetests/filetests/runtests/simd-fma-neg.clif), [`3`](https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/filetests/filetests/runtests/simd-fma.clif)). None of them test load sinking, so that would be a nice addition.

File: cranelift/codegen/src/isa/x64/lower.isle (hunk 1 above)
Comment (afonso360) [reply cranelift/codegen/src/isa/x64/lower.isle:2873]: > I forget what the details of our lowering and ISLE conversions are these days, but does this need to guard on sinkable loads?

We do some checks when converting from a `Value` into `XmmMem` where we force the value into a register if it's not eligible for being sunk (i.e. has multiple uses), so this looks okay to me.

**Review by afonso360 — APPROVED** [review]
This LGTM! I double checked the new instruction encodings against the intel manual and they all check out. Thanks!

File: cranelift/codegen/src/isa/x64/lower.isle (hunk 1 above)
Comment (fitzgen) [reply cranelift/codegen/src/isa/x64/lower.isle:2873]: Thanks for the clarification @afonso360 and @jameysharp!

**Comment by UnlimitedHummus** [comment]
To add runtime tests for load sinking, I would just duplicate some existing testcases and add loads to them. This doesn't test that the load is actually sunk, but rather that the results are still correct with load sinking. Is that the intention? Or am I missing some different type of testcase I could write?

**Comment by fitzgen** [comment]
That is a good thing to test, but generally I was just asking for runtests that check that the new lowerings that turn multiple clif instructions into a single machine instruction produce the same values that interpreting the individual clif instructions would.

See for example

https://github.com/bytecodealliance/wasmtime/blob/main/cranelift/filetests/filetests/runtests/iabs.clif#L1-L18

although we only care about the interpreter and x86-64 here.

File: cranelift/filetests/filetests/runtests/x64/fmsub.clif
Comment (UnlimitedHummus) [inline cranelift/filetests/filetests/runtests/x64/fmsub.clif:1]: Is this what you had in mind? I didn't find a better way to test this, other than to store the value on the stack and then load it from there again.

Comment (UnlimitedHummus) [reply cranelift/filetests/filetests/runtests/x64/fmsub.clif:1]: If that's good I will add some more tests of course.

Comment (afonso360) [reply cranelift/filetests/filetests/runtests/x64/fmsub.clif:1]: Yeah! Just something to ensure that the instruction gets emitted and runs properly. I wouldn't worry about trying to get all of the variations running, that might be too much (In my opinion).

File: cranelift/filetests/filetests/isa/x64/fmsub-inst.clif [hunk 2]
```diff
@@ -478,3 +478,242 @@ block0(v0: f64, v1: f64, v2: f64):
...
+; block1: ; offset 0x8
+;   leaq (%rsp), %r8
+;   vmovss %xmm1, (%r8) ; trap: heap_oob
+;   vfnmsub132ss (%r8), %xmm2, %xmm0 ; trap: heap_oob
+;   addq $0x10, %rsp
+;   movq %rbp, %rsp
+;   popq %rbp
+;   retq
+
+;; negate first, then load
```
Comment (UnlimitedHummus) [inline cranelift/filetests/filetests/isa/x64/fmsub-inst.clif:598]: This could generate more efficient core by selecting vfnmsub132ss instead and skipping the negation.

**Comment by UnlimitedHummus** [comment]
I added some tests, but also identified an optimization I missed. This only affects the case where the negation happens before the load. I think the order of the two shouldn't matter. I still need to look into how to write rules to handle that case.

**Comment by jameysharp** [comment]
> I added some tests, but also identified an optimization I missed. This only affects the case where the negation happens before the load. I think the order of the two shouldn't matter. I still need to look into how to write rules to handle that case.

This PR is already a nice improvement even if it doesn't optimize every case, so: If you want to dig into that you're welcome to, but I'd also be happy to merge this as-is. If you decide not to fix it right now, then it would be great to open a new issue describing what's not working, and include a link to that issue in the comments on the test that illustrates the problem.

That said, this is failing in CI right now. The error looks weird to me and I'm not sure whether you just need to re-run `CRANELIFT_TEST_BLESS=1 cargo test -p cranelift-tools --test filetests` or if there's something stranger going on, so please let us know if you have any trouble.

**Comment by UnlimitedHummus** [comment]
CI failing was due to me annotating a copy of the expected output. I moved my comment before the testcase, which fixed that. I tried adding some rules to improve code generation for this case yesterday, but that didn't work. So I'd have to take a closer look at ISLE to see why it doesn't generate the right code. Not sure when I'll get around to that so I opened an issue describing the missed optimization like you suggested.
